src/Ocelot/RateLimit/RateLimitCore.cs

```diff
@@ -0,0 +1,123 @@
+                var entry = _counterHandler.Get(counterId);
+                if (entry.HasValue)
+                {
+                    // entry has not expired
+                    if (entry.Value.Timestamp + rule.PeriodTimespan.Value >= DateTime.UtcNow)
+                    {
+                        // increment request count
+                        var totalRequests = entry.Value.TotalRequests + 1;
+
+                        // deep copy
+                        counter = new RateLimitCounter
+                        {
+                            Timestamp = entry.Value.Timestamp,
+                            TotalRequests = totalRequests
+                        };
+                    }
+                }
+
+                // stores: id (string) - timestamp (datetime) - total_requests (long)
+                _counterHandler.Set(counterId, counter, rule.PeriodTimespan.Value);
+            }
+
+            return counter;
+        }
+
+        public RateLimitHeaders GetRateLimitHeaders(ClientRequestIdentity requestIdentity, RateLimitOptions option)
+        {
+            var rule = option.RateLimitRule;
+            var headers = new RateLimitHeaders();
+            var counterId = ComputeCounterKey(requestIdentity, option);
+            var entry = _counterHandler.Get(counterId);
+            if (entry.HasValue)
+            {
+                headers.Reset = (entry.Value.Timestamp + ConvertToTimeSpan(rule.Period)).ToUniversalTime().ToString("o", DateTimeFormatInfo.InvariantInfo);
+                headers.Limit = rule.Period;
+                headers.Remaining = (rule.Limit - entry.Value.TotalRequests).ToString();
+            }
+            else
+            {
+                headers.Reset = (DateTime.UtcNow + ConvertToTimeSpan(rule.Period)).ToUniversalTime().ToString("o", DateTimeFormatInfo.InvariantInfo);
+                headers.Limit = rule.Period;
+                headers.Remaining = rule.Limit.ToString();
+            }
+
+            return headers;
+            throw new NotImplementedException();
+        }
+
+        public string ComputeCounterKey(ClientRequestIdentity requestIdentity, RateLimitOptions option)
+        {
+            var key =  $"{option.RateLimitCounterPrefix}_{requestIdentity.ClientId}_{option.RateLimitRule.Period}_{requestIdentity.HttpVerb}_{requestIdentity.Path}";
+
+            var idBytes = System.Text.Encoding.UTF8.GetBytes(key);
+
+            byte[] hashBytes;
+
+            using (var algorithm = System.Security.Cryptography.SHA1.Create())
+            {
+                hashBytes = algorithm.ComputeHash(idBytes);
+            }
+
+            return BitConverter.ToString(hashBytes).Replace("-", string.Empty);
+        }
+
+        public string RetryAfterFrom(DateTime timestamp, RateLimitRule rule)
+        {
+            var secondsPast = Convert.ToInt32((DateTime.UtcNow - timestamp).TotalSeconds);
+            var retryAfter = Convert.ToInt32(rule.PeriodTimespan.Value.TotalSeconds);
```

Comment:
Maybe check this is not null before operation or change property so it is not nullable.